content/blog/2019-11-13-space-efficient-gc/index.md

```diff
@@ -0,0 +1,100 @@
+
+![001](./001.png)
+
+Currently, most compilers always guarantee adequate alignment. However, it is hard to determine the proper position of the heap. Thus, this paper introduces a much less ad hoc and more flexible technique to avoid pointer misidentification.
+
+
+# Their Techniques
+
+This paper's techniques has two steps: 
+
+### Regular Garbage Collection
+
+They ensure regular garbage collections, with at least a fast and initial one happening just after system startup before any allocation begins.
+
+### Invalid Pointer Record
+
+A record of invalid pointers found during a garbage collection that could be valid object addresses afterwards is kept as a **blacklist**. The figure below shows the modified marking algorithm with blacklisting. The only modification is in bold, which adds adjacent addresses to the blacklist.
+
+![002.png](./002.png)
+
+The garbage collector does not allocate objects at blacklisted address. This scheme would mostly blacklist addresses that correspond to long-lived data values before these values become false references as they are the data that could possibly cause garbage to be retained indefinitely. 
+
+### Implementation
+
+They implemented variants of this approach in some versions of [PCR](https://dl.acm.org/citation.cfm?id=74862) and other [garbage collectors](https://dl.acm.org/citation.cfm?id=52202). They both conservatively scan the stacks, registers, static data and the heap.
+
+Entire pages rather than individual addresses are blacklisted. In that way, the blacklist can be implemented as a bit array, indexed by page numbers. Hask table can be utilized for discontiguous heaps.
+
+### Evaluation
+
+They evaluate their techniques by runing **a** program on different machines using both statically and dynamically linked versions of C library. The program allocates 200 circular linked lists containing 100 Kbytes each. And the collector would retain the entire list if any data points to any of the 100,000 addresses corresponding to objects in the list. The results of **storage retention** with and without blacklisting are shown in the following table.
+
+![003.png](./003.png)
+
+Several observations can be made based on the table:
+
+1. Blacklisting is effective in nearly eliminating all accidental retention caused by garbage collector conservativism.
+2. The numbers in the table are approximate as the results are not completely reproducible. This is due to the fact that the scanned part of the address space is polluted with UNIX environment variables. So they are specified as ranges.
+3. If all interior pointers are considered valid, it would be difficult to allocate individual objects larger than about 100 Kbytes without violating the blacklist constraint, or requesting memory from the OS at a garbage collector specified location.
+4. According to the paper, blacklist can be easily incorpoated into a garbage collecting allocator at *almost no performance cost*.
+
+### Other Sources of Excess Retention
```

Comment:
FWIW, it is somewhat unconventional by 2019 standards that the paper continues after the evaluation is done. It might be worth explicitly acknowledging that, at least in our in-class presentation: "This is where the paper probably should end, but it turns out that there is more stuff in here!"